clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -0,0 +1,391 @@
+                    return PT2IR_CONV_ERROR_GET_PENDING_EVENT;
+                }
+
+                /* Use a sideband session to check if pt_event is an image switch event.
+                 * If so, change the image in 'pt_instr_decoder_' to the target image.
+                 */
+                image = nullptr;
+                errcode =
+                    pt_sb_event(pt_sb_session_, &image, &event, sizeof(event), stdout, 0);
+                if (errcode < 0) {
+                    dx_decoding_error(errcode, "handle sideband event error", insn.ip);
+                    return PT2IR_CONV_ERROR_HANDLE_SIDEBAND_EVENT;
+                }
+
+                /* If it is not an image switch event, the PT instruction decoder
+                 * will not switch their cached image.
+                 */
+                if (image == nullptr)
+                    continue;
+
+                errcode = pt_insn_set_image(pt_instr_decoder_, image);
+                if (errcode < 0) {
+                    dx_decoding_error(errcode, "set image error", insn.ip);
+                    return PT2IR_CONV_ERROR_SET_IMAGE;
+                }
+            }
+            if ((nextstatus & pts_eos) != 0)
+                break;
+
+            /* Decode PT raw trace to pt_insn. */
+            status = pt_insn_next(pt_instr_decoder_, &insn, sizeof(insn));
+            if (status < 0) {
+                dx_decoding_error(status, "get next instruction error", insn.ip);
+                return PT2IR_CONV_ERROR_DECODE_NEXT_INSTR;
+            }
+
+            pt_insn_list_.push_back(insn);
+
+            /* TODO i#5505: Use drdecode to decode insn(pt_insn) to instr_t. */
+        }
+    }
+    return PT2IR_CONV_SUCCESS;
+}
+
+uint64_t
+pt2ir_t::get_instr_count()
+{
+    return pt_insn_list_.size();
+}
+
+void
+pt2ir_t::print_instrs_to_stdout()
+{
+    void *dcontext = GLOBAL_DCONTEXT;
+    for (auto pt_instr : pt_insn_list_) {
+        disassemble_with_info(dcontext, pt_instr.raw, STDOUT, false, false);
+    }
+}
+
+bool
+pt2ir_t::load_pt_raw_file(IN std::string &path)
+{
+    /* Under C++11, there is no good solution to get the file size after using ifstream to
+     * open a file. Because we will not change the PT raw trace file during converting, we
+     * don't need to think about write-after-read. We get the file size from file stat
+     * first and then use ifstream to open and read the PT raw trace file.
+     * XXX: We may need to update Dynamorio to support C++17+. Then we can implement the
+     * following logic without opening the file twice.
+     */
+    errno = 0;
+    struct stat fstat;
+    if (stat(path.c_str(), &fstat) == -1) {
+        ERRMSG("Failed to get file size of trace file: %s: %d.\n", path.c_str(), errno);
+        return false;
+    }
+    pt_raw_buffer_size_ = static_cast<size_t>(fstat.st_size);
+
+    pt_raw_buffer_ = std::unique_ptr<unsigned char>(new uint8_t[pt_raw_buffer_size_]);
```

Comment:
I would expect the new to use `char` since that's the type of this var.

Reply:
Done.